torch/lib/c10d/FileStore.cpp

```diff
@@ -0,0 +1,284 @@
+#include "FileStore.hpp"
+
+#include <sys/file.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <sstream>
+#include <system_error>
+#include <thread>
+
+#define SYSASSERT(rv) \
+  if ((rv) < 0) throw std::system_error(errno, std::system_category());
+
+#define SYSCHECK(expr) {                                                \
+    errno = 0; auto ___output = (expr); (void)___output;                \
+    SYSASSERT(errno);                                                   \
+  }
+
+namespace c10d {
+
+namespace {
+
+template<typename F>
+typename std::result_of<F()>::type syscall(F fn) {
+  while (true) {
+    auto rv = fn();
+    if (rv == -1) {
+      if (errno == EINTR) {
+        continue;
+      }
+    }
+    return rv;
+  }
+}
+
+// For a comprehensive overview of file locking methods,
+// see: https://gavv.github.io/blog/file-locks/.
+// We stick to flock(2) here because we don't care about
+// locking byte ranges and don't want locks to be process-wide.
+
+// RAII wrapper around flock(2)
+class Lock {
+ public:
+  explicit Lock(int fd, int operation) : fd_(fd) {
+    flock(operation);
+  }
+
+  ~Lock() {
+    unlock();
+  }
+
+  Lock(const Lock& that) = delete;
+
+  Lock(Lock&& other) noexcept {
+    fd_ = other.fd_;
+    other.fd_ = -1;
+  }
+
+  void unlock() {
+    if (fd_ >= 0) {
+      flock(LOCK_UN);
+      fd_ = -1;
+    }
+  }
+
+ protected:
+  int fd_;
+
+  void flock(int operation) {
+    auto rv = syscall(std::bind(::flock, fd_, operation));
+    if (rv < 0) {
+      throw std::system_error(errno, std::generic_category(), "flock");
+    }
+  }
+};
+
+class File {
+ public:
+  explicit File(const std::string& path, int flags) {
+    fd_ = syscall(std::bind(::open, path.c_str(), flags, 0644));
+    if (fd_ < 0) {
+      throw std::system_error(
+          errno,
+          std::generic_category(),
+          "Unable to open " + path);
+    }
+  }
+
+  ~File() {
+    ::close(fd_);
+  }
+
+  Lock lockShared() {
+    return Lock(fd_, LOCK_SH);
+  }
+
+  Lock lockExclusive() {
+    return Lock(fd_, LOCK_EX);
+  }
```

Comment:
A thought I had today: C++14 has a concept of shared mutex, and has a `std::shared_lock`, `std::unique_lock` as RAII style locking structs, and basically call into the object using the standard API (`lock()`, `unlock()` for exclusive access, `lock_shared()`, `unlock_shared()` for shared access).

I'm not sure what's the strategy for C++14 in c10 (I've seen some quips), but if we're going to go with that, it might be nice to use the standard APIs for this.

Of course not a blocker for this PR.